std/path.d

```diff
@@ -1823,138 +1615,414 @@ unittest
+    {
+        assert (toNormalizedPath("/foo/./bar/..//baz/").array == "/foo/baz");
+        assert (toNormalizedPath("../foo/.").array == "../foo");
+        assert (toNormalizedPath("/foo/bar/baz/").array == "/foo/bar/baz");
+        assert (toNormalizedPath("foo/./bar/../../", "../baz").array == "../baz");
+        assert (toNormalizedPath("/foo/./bar/../../baz").array == "/baz");
+    }
+
+    version (Windows)
+    {
+        assert (toNormalizedPath(`c:\foo\.\bar/..\\baz\`).array == `c:\foo\baz`);
+        assert (toNormalizedPath(`..\foo\.`).array == `..\foo`);
+        assert (toNormalizedPath(`c:\foo\bar\baz\`).array == `c:\foo\bar\baz`);
+        assert (toNormalizedPath(`c:\foo\bar/..`).array == `c:\foo`);
+        assert (toNormalizedPath(`\\server\share\foo\..\bar`).array == `\\server\share\bar`);
+    }
+    ---
+*/
+
+auto toNormalizedPath(R)(R path)
+        if (isSomeChar!(ElementEncodingType!R) &&
+            (isRandomAccessRange!R && hasSlicing!R && hasLength!R || isNarrowString!R))
+{
+    alias C = Unqual!(ElementEncodingType!R);
+    alias S = typeof(path[0..0]);
+
+    static struct Result
+    {
+        @property bool empty()
+        {
+            return c == c.init;
+        }
+
+        @property C front()
+        {
+            return c;
+        }
+
+        void popFront()
+        {
+            C lastc = c;
+            c = c.init;
+            if (element.empty)
+            {
+              L1:
+                while (1)
+                {
+                    if (elements.empty)
+                    {
+                        element = element[0..0];
+                        return;
+                    }
+                    element = elements.front;
+                    elements.popFront();
+                    if (isDot(element) || (rooted && isDotDot(element)))
+                        continue;
+                    break;
+                }
+
+                if (rooted || !isDotDot(element))
+                {
+                    int n = 1;
+                    auto elements2 = elements.save;
+                    while (!elements2.empty)
+                    {
+                        auto e = elements2.front;
+                        elements2.popFront();
+                        if (isDot(e))
+                            continue;
+                        if (isDotDot(e))
+                        {
+                            --n;
+                            if (n == 0)
+                            {
+                                elements = elements2;
+                                element = element[0..0];
+                                goto L1;
+                            }
+                        }
+                        else
+                            ++n;
+                    }
+                }
+
+                if (lastc == dirSeparator[0] || lastc == lastc.init)
+                    c = getElement0();
+                else
+                    c = dirSeparator[0];
```

Comment:
`static assert(dirSeparator.length == 1);` otherwise the whole approach needs rethinking.